contributors/design-proposals/scheduling/coscheduling.md

```diff
@@ -0,0 +1,174 @@
+  - "@k82cn"
+owning-sig: sig-scheduling, machine-learning WG
+reviewers:
+  - "@bsalamat"
+  - "@vishh"
+approvers:
+  - "@bsalamat"
+  - "@vishh"
+editor: TBD
+creation-date: 2018-07-03
+last-updated: 2018-09-06
+status: provisional
+---
+
+# Coscheduling
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Motivation](#motivation)
+* [Function Detail](#function%20detail)
+  * [API Definitation](#api%20definition)
+  * [Lifecycle Management](#lifecycle%20management)
+  * [kube-arbitrator](#kube-arbitrator)
+  * [Customized Controller](#customized%20controller)
+* [Feature Interaction](#feature%20interaction)
+  * [Multi-scheduler](#multi-scheduler)
+  * [Priority/Preemption](#priority/preemption)
+  * [Pod RestartPolicy](#pod%20restartPolicy)
+  * [Admission Controller](#admission%20controller)
+  * [Kubectl](#kubectl)
+* [References](#references)
+
+## Motivation
+
+After the discussion at [Coscheduling](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#heading=h.ckn8nv2jj0xv) proposal, we decide to implement Coscheduling in [kube-arbitrator](https://github.com/kubernetes-incubator/kube-arbitrator) by CRDs. kube-arbitrator focuses on "batch" workload in kubernetes, and will share the same [scheduling frameworks](https://github.com/kubernetes/community/pull/2281) when it's ready. This document is used to provide definition of API object and the scheduler behaviour of Coscheduling.
+
+## Function Detail
+
+### API Definition
+
+The following requirements are identified during the discussion of this feature:
+
+1. Existing workload can use this feature without (or with a few) configuration changes
+2. Pods of a group/gang may have different `PodSpec` (and/or belong to different collections)
+3. Existing controllers which are responsible for managing life cycle of collections work well with this feature
+
+To meet the requirements above, the following **Kind** is introduced in core part under `scheduling.k8s.io/v1alpha1` **Group**/**Version**.
+
+```go
+// PodGroup defines the scheduling requirement of a pod group
+type PodGroup struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+	// Spec defines the behavior of a pod group.
+	// +optional
+    Spec PodGroupSpec
+
+	// Status represents the current information about a pod group.
+	// This data may not be up to date.
+	// +optional
+    Status PodGroupStatus
+}
+
+// PodGroupSpec represents the template of a pod group.
+type PodGroupSpec struct {
+    // NumMember defines the number of members/tasks to run the pod group;
+    // if there's not enough resources to start all tasks, the scheduler
+    // will not start anyone.
+    NumMember int
+
+    // Resources defines the total resource the PodGroup requests to run
+    // Pods.
+    Resources v1.ResourceList
+}
+
+// PodGroupStatus represents the current state of a pod group.
+type PodGroupStatus struct {
+    // The number of actively running pods.
+    // +optional
+    Running int32
```

Comment:
Can we add a list of `Admitted` pods as well? Rationale is that once a pod group resource quota request is granted, we want to know how many pods have already been admitted as part of the group. 

Also, would it make sense to have a list of local object references to identify pods rather than just an integer count. That way it will be easier for higher level controllers to track the phases of pods belonging to a pod group easily.

Reply:
> Also, would it make sense to have a list of local object references to identify pods rather than just an integer count.

Wanted to add such detail for user; but I think the customized controller will provide such info. Prefer to keep it simple right now, if it's necessary, let's update it in other PRs. That's the benefits we got from CRD :)